core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -0,0 +1,379 @@
+    * to. The store then must choose from any of the provided media types. Any methods that are invoked after that
+    * accept a key argument will be guaranteed to have this key be in the desired media type. Also any return values
+    * from this store should be in this media type
+    * <p>
+    * Note that choosing a media type other than the storage media type will incur the cost of converting the key
+    * to and from the storage media type for input and output parameters.
+    * @param storageMediaType how the key is stored in memory
+    * @param supportedMediaTypes what media types Infinispan can convert to automatically on behalf of the store
+    * @return the media type the store desires for keys to be in when invoked and will return in
+    */
+   default MediaType getKeyMediaType(MediaType storageMediaType, Set<MediaType> supportedMediaTypes) {
+      return storageMediaType;
+   }
+
+   /**
+    *
+    * @param storageMediaType
+    * @param supportedMediaTypes
+    * @return
+    */
+   default MediaType getValueMediaType(MediaType storageMediaType, Set<MediaType> supportedMediaTypes) {
+      return storageMediaType;
+   }
+
+   /**
+    *
+    * @return
+    */
+   default CompletionStage<Boolean> isAvailable() {
+      return CompletableFutures.completedTrue();
+   }
+
+   /**
+    *
+    * @param segment
+    * @param key
+    * @return
+    */
+   CompletionStage<MarshallableEntry<K, V>> loadEntry(int segment, Object key);
+
+   /**
+    *
+    * @param segment
+    * @param key
+    * @return
+    */
+   default CompletionStage<Boolean> containsKey(int segment, Object key) {
+      return loadEntry(segment, key)
+            .thenApply(Objects::nonNull);
+   }
+
+   /**
+    *
+    * @param segment
+    * @param entry
+    * @return
+    */
+   CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> entry);
+
+   /**
+    *
+    * @param segment
+    * @param key
+    * @return
+    */
+   CompletionStage<Boolean> delete(int segment, Object key);
+
+   /**
+    *
+    * @param segments
+    * @return
+    */
+   default CompletionStage<Void> addSegments(IntSet segments) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.SEGMENTABLE + ", but it does not implement addSegments");
+   }
+
+   /**
+    *
+    * @param segments
+    * @return
+    */
+   default CompletionStage<Void> removeSegments(IntSet segments) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.SEGMENTABLE + ", but it does not implement removeSegments");
+   }
+
+   /**
+    *
+    * @return
+    */
+   CompletionStage<Void> clear();
+
+   /**
+    *
+    * @param publisherCount
+    * @param publisher
+    * @return
+    */
+   default CompletionStage<Void> bulkWrite(int publisherCount, Publisher<SegmentedPublisher<MarshallableEntry<K, V>>> publisher) {
```

Comment:
add a segment to `MarshallableEntry`?

Reply:
That is a decent alternative. I will have to think more about that.

One thing the SegmentedPublisher interface does is group together entries of the same segment. This can be good or bad depending on if the store can optimize for grouped segments or not. And if we ever change putAll to already group similar segment values, it will be able to gain benefits from that.

On the other hand having a single Publisher type would be simpler for external users.

Reply:
I've never played with Publishers... I just found it weird to have a Publisher of SegmentedPublisher. 
From what I can see, isn't possible to do any grouping with the Publisher, right?

Reply:
Hrmm, RxJava has support for grouping if that is what you are referring to. http://reactivex.io/RxJava/3.x/javadoc/io/reactivex/rxjava3/core/Flowable.html#groupBy-io.reactivex.rxjava3.functions.Function-
Our SegmentPublisher would be just like the GroupedFlowable.

Reply:
As a side note, if a cache is not SEGMENTABLE this will return only a single SegmentedPublisher instead of breaking them into segments.

Reply:
ok. your solution is better than messing up with `MarshallableEntry`.